src/s3/args.rs

```diff
@@ -1112,70 +1206,64 @@ impl<'a> ListObjectVersionsArgs<'a> {
     /// Specifies the bucket name on which listing is to be performed.
-    pub bucket: &'a str,
+    pub bucket: String,
     /// Delimiter to roll up common prefixes on.
-    pub delimiter: Option<&'a str>,
-    pub use_url_encoding_type: bool,
+    pub delimiter: Option<String>,
+    pub encoding_type: EncodingType,
+    pub max_keys: Option<u16>,
+    pub prefix: Option<String>,
+
     /// Used only with ListObjectsV1.
-    pub marker: Option<&'a str>,
+    pub marker: Option<String>,
+
     /// Used only with ListObjectsV2
-    pub start_after: Option<&'a str>,
-    /// Used only with GetObjectVersions.
-    pub key_marker: Option<&'a str>,
-    pub max_keys: Option<u16>,
-    pub prefix: Option<&'a str>,
+    pub start_after: Option<String>,
     /// Used only with ListObjectsV2.
-    pub continuation_token: Option<&'a str>,
+    pub continuation_token: Option<String>,
     /// Used only with ListObjectsV2.
     pub fetch_owner: bool,
-    /// Used only with GetObjectVersions.
-    pub version_id_marker: Option<&'a str>,
     /// MinIO extension for ListObjectsV2.
     pub include_user_metadata: bool,
+
+    /// Used only with GetObjectVersions.
+    pub key_marker: Option<String>,
+    /// Used only with GetObjectVersions.
+    pub version_id_marker: Option<String>,
+
+    /// This parameter takes effect only when delimiter is None. Enables
+    /// recursive traversal for listing of the bucket and prefix.
     pub recursive: bool,
     /// Set this to use ListObjectsV1. Defaults to false.
     pub use_api_v1: bool,
     /// Set this to include versions.
     pub include_versions: bool,
-    /// A callback function to process results of object listing.
-    pub result_fn: &'a dyn Fn(Vec<Item>) -> bool,
 }
 
-impl<'a> ListObjectsArgs<'a> {
+impl ListObjectsArgs {
     /// Returns argument for [list_objects()](crate::s3::client::Client::list_objects) API with given bucket name and callback function for results.
     ///
     /// # Examples
     ///
     /// ```
     /// use minio::s3::args::*;
-    /// let args = ListObjectsArgs::new(
-    ///     "my-bucket",
-    ///     &|items| {
-    ///         for item in items.iter() {
-    ///             println!("{:?}", item.name);
-    ///         }
-    ///         true
-    ///     },
-    /// ).unwrap();
+    /// let args = ListObjectsArgs::new("my-bucket").unwrap();
     /// ```
-    pub fn new(
-        bucket_name: &'a str,
-        result_fn: &'a dyn Fn(Vec<Item>) -> bool,
-    ) -> Result<ListObjectsArgs<'a>, Error> {
+    pub fn new(bucket_name: &str) -> Result<ListObjectsArgs, Error> {
         check_bucket_name(bucket_name, true)?;
 
         Ok(ListObjectsArgs {
             extra_headers: None,
             extra_query_params: None,
             region: None,
-            bucket: bucket_name,
+            bucket: bucket_name.to_owned(),
             delimiter: None,
-            use_url_encoding_type: true,
+            encoding_type: EncodingType::default(),
```

Comment:
There is no need to have type for encoding. For any `EncodingType`, we handle only one kind which is `url` encoding i.e. if encoding type is `url`, we return decoded values else we return as is.

Having `use_url_encoding_type bool` is good enough for all cases.

Reply:
With `use_url_encoding_type` we are making it opt-in to use the encoding parameter. I wanted to make it the default because the purpose of the parameter is to handle encoding unicode characters in object names that cannot be represented in the xml response - quoted from https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListObjects.html


> [encoding-type](https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListObjects.html#API_ListObjects_RequestSyntax)
> 
> Requests Amazon S3 to encode the object keys in the response and specifies the encoding method to use. An object key can contain any Unicode character; however, the XML 1.0 parser cannot parse some characters, such as characters with an ASCII value from 0 to 10. For characters that are not supported in XML 1.0, you can add this parameter to request that Amazon S3 encode the keys in the response.
> 
> Valid Values: url
> 

This is a good reason to always set `encoding-type=url` (i.e. default setting) because user will not run into this encoding issue mentioned.

I used a type for it to allow for future changes - in case a new encoding type is added, it would be easy to fix the library with a minor update (instead of a breaking change). If a new encoding type `foo` is added in the future, anyone can start using it as just `EncodingType::Other("foo")`, and we can update the library adding `EncodingType::Foo`. With a bool we just lose all these options.

If you think it's overkill I'll change it to `use_url_encoding_type` and set it to `true` by default.

Reply:
`EncodingType` is useful if we handle more than one encoding type. Practically it is either `url` type or none. `use_url_encoding_type` tells what `minio-rs` does with object list xml response. We could keep the current behavior i.e. `use_url_encoding_type = true` as default.

Reply:
Done @balamurugana